src/domain/datasets/domain/src/use_cases/update_version_file_use_case.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use file_utils::MediaType;
+use internal_error::InternalError;
+use odf::dataset::RefCASError;
+use thiserror::Error;
+use tokio::io::AsyncRead;
+
+use crate::{ExtraDataFields, FileVersion};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait UpdateVersionFileUseCase: Send + Sync {
+    async fn execute(
+        &self,
+        dataset_handle: &odf::DatasetHandle,
+        content_args_maybe: Option<ContentArgs>,
+        expected_head: Option<odf::Multihash>,
+        extra_data: Option<ExtraDataFields>,
+    ) -> Result<UpdateVersionFileResult, UpdateVersionFileUseCaseError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct ContentArgs {
+    pub content_length: usize,
+    pub content_hash: odf::Multihash,
```

Comment:
Not an expert in the topic, but maybe we could avoid double scanning of the content, if this hash was optional?